File: src/NovelRT/Ink/InkService.cpp
// Copyright © Matt Jones and Contributors. Licensed under the MIT Licence (MIT). See LICENCE.md in the repository root
// for more information.
#if defined(__clang__)
#pragma clang diagnostic push
#pragma clang diagnostic ignored "-Wunused-lambda-capture"
#endif

#include <NovelRT.h>

using namespace NovelRT::Utilities;

namespace NovelRT::Ink
{
    InkService::InkService(std::shared_ptr<DotNet::RuntimeService> runtimeService,
                           void (*getExports)(Exports* exports)) noexcept
        : _runtimeService(runtimeService), _exports(Lazy<Exports>([getExports, this] {
              Exports exports;
              getExports(&exports);
              return exports;
          }))
    {
    }

    void InkService::initialise()
    {
        _exports.getActual().Initialise();
    }

    void InkService::tearDown()
    {
        if (_exports.isCreated())
        {
            _exports.getActual().Teardown();
        }
    }

    std::shared_ptr<Story> InkService::createStory(const char* jsonString)
    {
        return std::make_shared<Story>(shared_from_this(), _exports.getActual().GetStoryExports, jsonString);
    }

    std::shared_ptr<DotNet::RuntimeService> InkService::getRuntimeService()
    {
        return _runtimeService;
    }
} // namespace NovelRT::Ink

#if defined(__clang__)
#pragma clang diagnostic pop
#endif

File: src/NovelRT.DotNet/InkService.cs
// Copyright © Matt Jones and Contributors. Licensed under the MIT Licence (MIT). See LICENCE.md in the repository root for more information.

using System;
using System.Runtime.InteropServices;
using Ink.Runtime;

namespace NovelRT.DotNet
{
    public static unsafe partial class InkService
    {
        [UnmanagedCallersOnly]
        public static void GetExports(Exports* exports)
        {
            exports->Initialise = &Initialise;
            exports->Teardown = &Teardown;
            exports->GetStoryExports = &GetStoryExports;
        }

        [UnmanagedCallersOnly]
        private static void Initialise()
        {
            #if DEBUG
            Console.WriteLine("Initialising the .NET ink service...");
            #endif
        }

        [UnmanagedCallersOnly]
        private static void GetStoryExports(StoryExports* exports)
        {
            exports->CreateFromJsonString = &Story_CreateFromJsonString;
            exports->ChooseChoiceIndex = &Story_ChooseChoiceIndex;
            exports->Continue = &Story_Continue;
            exports->ContinueMaximally = &Story_ContinueMaximally;
            exports->GetCanContinue = &Story_GetCanContinue;
            exports->ResetState = &Story_ResetState;
        }

        [UnmanagedCallersOnly]
        private static IntPtr Story_CreateFromJsonString(byte* jsonString)
        {
            var story = new Story(RuntimeService.PointerToString(jsonString));
            return RuntimeService.AllocateHandle(story);
        }

        [UnmanagedCallersOnly]
        private static void Story_ChooseChoiceIndex(IntPtr storyHandle, int choiceIdx)
        {
            var story = RuntimeService.ResolveHandle<Story>(storyHandle);
            story.ChooseChoiceIndex(choiceIdx);
        }

        [UnmanagedCallersOnly]
        private static byte* Story_Continue(IntPtr storyHandle)
        {
            var story = RuntimeService.ResolveHandle<Story>(storyHandle);
            return RuntimeService.StringToPointer(story.Continue());
        }

        [UnmanagedCallersOnly]
        private static byte* Story_ContinueMaximally(IntPtr storyHandle)
        {
            var story = RuntimeService.ResolveHandle<Story>(storyHandle);
            return RuntimeService.StringToPointer(story.ContinueMaximally());
        }

        [UnmanagedCallersOnly]
        private static byte Story_GetCanContinue(IntPtr storyHandle)
        {
            var story = RuntimeService.ResolveHandle<Story>(storyHandle);
            return story.canContinue ? (byte)1 : (byte)0;
        }

        [UnmanagedCallersOnly]
        private static void Story_ResetState(IntPtr storyHandle)
        {
            var story = RuntimeService.ResolveHandle<Story>(storyHandle);
            story.ResetState();
        }

        [UnmanagedCallersOnly]
        private static void Teardown()
        {
            #if DEBUG
            Console.WriteLine("Tearing down the .NET ink service...");
            #endif
        }
    }
}
